alicat/util.py

```diff
@@ -108,123 +136,67 @@ class TcpClient(Client):
-
-        As industrial devices are commonly unplugged, this has been expanded to
-        handle recovering from disconnects.
-        """
-        await self._handle_connection()
-        self.writer.write(message.encode() + self.eol)
-
-    async def _handle_connection(self) -> None:
-        """Automatically maintain TCP connection."""
-        if self.open:
-            return
-        try:
-            await asyncio.wait_for(self._connect(), timeout=0.75)
-            self.reconnecting = False
-        except (asyncio.TimeoutError, OSError):
-            if not self.reconnecting:
-                logger.error(f'Connecting to {self.address} timed out.')
-            self.reconnecting = True
-
-    async def _handle_communication(self, command: str) -> str | None:
-        """Manage communication, including timeouts and logging."""
-        try:
-            await self._write(command)
-            future = self._readline()
-            result = await asyncio.wait_for(future, timeout=0.75)
-            self.timeouts = 0
-            return result
-        except (asyncio.TimeoutError, TypeError, OSError):
-            self.timeouts += 1
-            if self.timeouts == self.max_timeouts:
-                logger.error(f'Reading from {self.address} timed out '
-                             f'{self.timeouts} times.')
-                await self.close()
-            return None
+        return reader, writer
 
 
 class SerialClient(Client):
     """Client using a directly-connected RS232 serial device."""
 
-    def __init__(self, address: str, baudrate: int=19200, timeout: float=.15):
-        """Initialize serial port."""
+    baudrate: int
+    bytesize: int
+    stopbits: int
+    parity: str
+
+    def __init__(self,
+                 address: str,
+                 baudrate: int = 19200,
+                 timeout: float = 10.0,
+                 bytesize: int = serial.EIGHTBITS,
+                 stopbits: int = serial.STOPBITS_ONE,
+                 parity: str = serial.PARITY_NONE):
+
         super().__init__(timeout)
+
         self.address = address
         assert isinstance(self.address, str)
+
         self.baudrate = baudrate
-        self.timeout = timeout
-        self.connectTask = asyncio.create_task(self._connect())
+        self.bytesize = bytesize
+        self.stopbits = stopbits
+        self.parity = parity
 
+    async def _connect(self) -> tuple[asyncio.StreamReader, asyncio.StreamWriter]:
+        """Asynchronously open a TCP connection with the server."""
+        await self.close()
 
-    async def _connect(self) -> None:
-        self.reader, self.writer = await serial_asyncio_fast.open_serial_connection(
+        reader, writer = await serial_asyncio_fast.open_serial_connection(
             url = self.address,
             baudrate = self.baudrate,
-            bytesize = serial.EIGHTBITS,
-            stopbits = serial.STOPBITS_ONE,
-            parity = serial.PARITY_NONE,
+            bytesize = self.bytesize,
```

Comment:
I think when you did get `git merge master` you un-did some of the changes I had made in `master` when resolving conflicts. 

https://github.com/alexrudd2/alicat/commit/f6f0166f35e88b6914172f66a6e28395f3590adf
https://github.com/alexrudd2/alicat/commit/36b8850ff971939c147cdd25bb8d57554e0ab938

I'm trying to merge in portions we're both happy with so as to shrink the diff over time.  But it does make merge conflicts a bit trickier to resolve - sorry about that.

Reply:
Many apologies -- I thought I handled the merge appropriately.

Do I understand correctly that you're handling this, or is this a request for me to jump in and re-add those changes?

Reply:
Either way is fine for me.  If you have bandwidth make the changes, great.   Otherwise, I will cherry-pick the changes from this PR in a new branch.

The only thing I don't want to merge right now is changing the connection logic.  That may well come later - there have been plenty of breaking changes already!   But it should be a single decision (easier to discuss and/or revert) rather than lumped in with others.

Reply:
Sounds good.  I'm a bit low on bandwidth right now, and my fork is working well for my project, so I'm happy.  Feel free to take this at your own pace.  I think the most important bits are just the lock and timeouts -- feel free to toss everything else.